New feature: Add username check , Issue #9867

framontb opened the pull request:
Pull Request for Issue #9867 .
#### Summary of Changes

Changes to add checking of 'username' field in registration form.
#### Testing Instructions
- Configure usernamecheck settings in back end (User -> options): minimum number of chars, maximum number of chars, character set.
- Allow User Registration in Front End (Users -> Options ->  Allow User Registration)
- Try to register with usernamecheck non-compliance requirements for username, and with compliance requirements.
#### Additional information

This feature started as a plugin but soon it change to add the feature to the com_users backend API for the sake of usability.

framontb committed:
New feature: plugin plg_user_usernamecheck , Issue #9867

ggppdk commented:
**Implementation looks very good**

just some comments about UI and placemennt of code (**please see pictures**)

- there is already a joomla user plugin that can host the event method **onUserBeforeSave**
this is: **plugins/user/joomla/joomla.xml**

thus adding a new plugin just to add this 1 method can be avoided

- there are several parameters for password limitations in the parameters  of  USERs  component (com_users) and these parameters are placed inside the plugin, that is somewhat an inconsistency for the UI

Plus it is always more difficult for users to find parameters in so many plugins ... that a web-site has ... instead the user would expect to find the parameters in the "Users manager" where the password parameters already exist.

Please see the pictures that are the result of modifying the  **config.xml** of the users component

**For spacers with "label" class**
```
<field type="spacer" name="pass_lims" label="Password limitations" class="label" />
...
<field type="spacer" name="uname_lims" label="Username limitations" class="label" />
```

**OR for tabs**
```
</fieldset><fieldset name="pass_lims" label="Password" >
...
</fieldset><fieldset name="user_lims" label="Username" >
```

![username_lims1](https://cloud.githubusercontent.com/assets/5092940/14579262/2946f00e-03ab-11e6-97ad-9211a5360311.png)
![username_lims2](https://cloud.githubusercontent.com/assets/5092940/14579263/2f33a61a-03ab-11e6-8b19-504b87558874.png)
![username_lims3](https://cloud.githubusercontent.com/assets/5092940/14579265/31ecc9f4-03ab-11e6-9d3b-386fa028e8c1.png)

bembelimen commented on plugins/user/usernamecheck/usernamecheck.php:
```diff
@@ -0,0 +1,125 @@
+	{
+		// If we aren't saving a "new" user (registration), or if we are not 
+		// in the front end of the site, the let the save happen withour interrption.
+		if (!$isNew || !JFactory::getApplication()->isSite()) {
```
What happens, if a user edits its own username? Shouldn't the validation trigger evertime?

framontb replied:
It seems not the best way to place this functionality in a plugin, so I'll fix the bug but in a third party plugin.
The same for the rest or the comments (so many and so good !).
Thanks !

bembelimen commented on plugins/user/usernamecheck/usernamecheck.php:
```diff
@@ -0,0 +1,125 @@
+		}
+		
+		// Load the language file for the plugin
+		$this->loadLanguage();
```
```->loadLanguage()``` can be initialized by a class variable

bembelimen commented on plugins/user/usernamecheck/usernamecheck.php:
```diff
@@ -0,0 +1,125 @@
+		
+		// Create the input object
+		// http://joomla.stackexchange.com/questions/8883/joomla-3-3-deprecated-function-for-jrequestgetvar
+		$input = JFactory::getApplication()->input;
```
```JFactory::getApplication()``` can be loaded via ```$this->app``` (if you set the class variable)

bembelimen commented on plugins/user/usernamecheck/usernamecheck.php:
```diff
@@ -0,0 +1,125 @@
+		// Only check if minNumChars field is set
+		if ($minNumChars != 0){
+			// Check if $usernameLenght achieve minimum lenght
+			if ($usernameLenght < $minNumChars) {
```
Join the two ifs
```
$minNumChars = (int) $this->params->get('minNumChars');

if ($minNumChars && $usernameLenght < $minNumChars) {

bembelimen commented on plugins/user/usernamecheck/usernamecheck.php:
```diff
@@ -0,0 +1,125 @@
+		$input = JFactory::getApplication()->input;
+		
+		// Get the $username field
+		$username = $data['username'];
```
Why this extra assignment? Just use ```$data['username']``` the whole time.

bembelimen commented on plugins/user/usernamecheck/usernamecheck.php:
```diff
@@ -0,0 +1,125 @@
+		// Only check if maxNumChars field is set
+		if ($maxNumChars != 0){
+			// Check if $usernameLenght achieve minimum lenght
+			if ($usernameLenght > $maxNumChars) {
```
Join here, too. + see comment above

bembelimen commented on plugins/user/usernamecheck/usernamecheck.php:
```diff
@@ -0,0 +1,125 @@
+		$charset = $this->params->get('charset');
+		
+		// Only check if $charset field is set
+		if (!empty($charset)) {
```
You could fasten this up with:
```
$charset = str_split($this->params->get('charset'));
if ($data['username'] != str_replace($charset, '', $data['username']))
{
    // Error message
}
```

framontb replied:
I fix all except this one, because I want to notify the user what were the wrong characters.

bembelimen replied:
$charset = str_split($this->params->get('charset'));
$uname = str_split($data['username']);

$invalid_chars = array_intersect($charset, $uname);

if (!empty($invalid_chars)) { echo 'Your invalid chars are: ' . implode(',', $invalid_chars); }

framontb replied:
Great point ! Idid'n even know this "array_intersect()" command...
Finally I fix this too (in the plugin by now: https://github.com/framontb/plg_user_usernamecheck), but actually, instead of using the ''array_intersect" command I use the "array_diff".
''array_intersect" gives me the VALID chars instead of the INVALID chars that I get with "array_diff". 
Furthermore I use the StringHelper::str_split() wrapper for UTF-8 compatibility, and the "array_unique()" command for not repeating chars.
Thanks a lot !

bembelimen replied:
Ahh, yes, charset holds the valid characters, true, diff is the correct solution

chrisdavenport commented:
I think this might be a good opportunity to incorporate Unicode abuse checking too.  For example, see

http://www.m3aawg.org/sites/default/files/m3aawg-unicode-tutorial-2016-02.pdf
http://www.m3aawg.org/sites/default/files/m3aawg-unicode-best-practices-2016-02.pdf

framontb commented:
OK @ggppdk, all you said looks good to me, and have all the sense. If no one have more objections, I'll place the code just inside the USERs component to achieve the UI you proposes in the picture. 
I'm going to read the "Unicode abuse" too, trying to find out what's going on there.
I think, I'm going to publish the plugin 'usernamecheck' as a third party extension, so people (@baijianpeng) can apply usernamecheck plugin right now in 3.5 versions. Is this right way?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9943">issues.joomla.org/joomla-cms/9943</a>.</sub>

baijianpeng commented:
Thank you @framontb , I prefer to see this feature as a Joomla core
com_users feature. In this way, it will help more users like me.



2016-04-16 20:35 GMT+08:00 framontb <notifications@github.com>:

> OK @ggppdk <https://github.com/ggppdk>, all you said looks good to me,
> and have all the sense. If no one have more objections, I'll place the code
> just inside the USERs component to achieve the UI you proposes in the
> picture.
> I'm going to read the "Unicode abuse" too, trying to find out what's going
> on there.
> I think, I'm going to publish the plugin 'usernamecheck' as a third party
> extension, so people (@baijianpeng <https://github.com/baijianpeng>) can
> apply usernamecheck plugin right now in 3.5 versions. Is this right way?
> ------------------------------
> This comment was created with the J!Tracker Application
> <https://github.com/joomla/jissues> at issues.joomla.org/joomla-cms/9943
> <https://issues.joomla.org/tracker/joomla-cms/9943>.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/joomla/joomla-cms/pull/9943#issuecomment-210805946>
>

framontb commented:
Anyway: https://github.com/framontb/plg_user_usernamecheck

andrepereiradasilva commented:
> For me to test this plugin was necessary to discover and install the usernamecheck plugin prior to test. I suppose, this is not the right way, because core plugins should be installed itself. I don't know how to do this.

- You need to add then to the extension table for the 3 database types supported: mysql, postgresql, sqlazure
 Example for mysql (plg_system_stats): https://github.com/joomla/joomla-cms/blob/staging/installation/sql/mysql/joomla.sql#L613

- Also you need to add on update sql (for the 3 database types too)
 Example for mysql (plg_system_stats): https://github.com/joomla/joomla-cms/blob/staging/administrator/components/com_admin/sql/updates/mysql/3.5.0-2015-11-05.sql#L1-L2

- And also need to add it to the update manifest list on update
 Example for plg_system_stats: https://github.com/joomla/joomla-cms/blob/staging/administrator/components/com_admin/script.php#L343

- Also i think you need to add the language files to the install.xml file
 Example for plg_system_stats: https://github.com/joomla/joomla-cms/blob/staging/administrator/language/en-GB/install.xml#L204-L205

Don't know if anything else is needed

framontb committed:
Revert "New feature: plugin plg_user_usernamecheck , Issue #9867"

This reverts commit e9a547ea898b96515c483ae70de6301e9542bac3.

ggppdk commented:
@framontb

Actually to move the code inside users component is **rather little work**. The place to add the code is already there, the username validation code is supposed to be inside file:

**joomla/form/rule/username.php**

1. Place usernane parameters (as described above) inside XML file:
administrator/components/com_users/config.xml

2. inside the **test()** function of **joomla/form/rule/username.php**:
**$params = JComponentHelper::getParams('com_users');**

3. return false / true, if the username is acceptable
currently the existing code only check for duplicate name

About limitations of **password**, you will see that they are being checked inside
**joomla/cms/rule/username.php**, also calling: $params = JComponentHelper::getParams('com_users');


But 2 notes the registration form:
Currently the config.xml is using **type="text"** for type of the **username**

- [ ] it would be nice if someone added an AJAX validation type, i say AJAX validation because
e.g. the duplicate username check can not be handled from the form:
```
name="username" type="text" validate="ajax" validate_url="..." 
// or
name="username" type="ajax_text" validate_url="..." 
```
- [] Registration form needs some inline message or something about the limitations of the username
maybe adding a new form element  "username" ??
exactly like this exists:
libraries\joomla\form\fields\password.php
there should be a:
libraries\joomla\form\fields\username.php

smz commented:
I guess the relevant strings (such as JLIB_DATABASE_ERROR_VALID_AZ09) should also be modified...

framontb committed:
Adding a "Username options" tab in the administrative options of com_users (config.xml)
Adding three options (config.xml):
	- MINimum number of characters
	- MAXimum number of characters
	- Character set ALLOWED
Adding translations (en-GB.com_users.ini)

framontb commented:
Don't merge !
I'm still placing code.

framontb committed:
Moving some error messages to front end.
Placing the test code in joomla/form/rule/username.php.
Changing field names and types.

framontb commented:
@ggppdk 

I think what you suggested in your post is done (not the AJAX validation). You were right, it was straightforward putting the code there (joomla/form/rule/username.php).
I have two more questions:
- Should I add the password tab here in this issue?
- How could I set a value for the fields in username? (I try with "value", but doesn't work.)
Thanks in advance for you invaluable help.

ggppdk commented:
If "Username" (limitations) gets a different TAB 
then it makes sense to have seperate TAB for "Password" too
or maybe both should be in the 1 new TAB calling it "Validation",
- anyway THIS IS NOT MY CALL, someone else should say about this, everything i wrote here are just opinions, **you can just do it** and then see if maintainers agree with it, i am just a user here


Now about your question of how to "set a value", 
- i think you mean **how to use "default" values**, you need to do **2 things** 

1. almost all "form elements" (parameters) used in XML file respect / use the attribute  "default"
e.g. <field name="maximum_length_username"  ... **default="99"**  ...

2.  then inside your code use the **same default value**:
$maxNumChars = $params->get('maximum_length_username', 99);

-

Some **notes / concerns**:

-

- [ ] Default values of new features should be wide enough

**not to change / break existing behaviour** in existing installations, e.g. you added a default value of max characters 99 for username, this is good should not cause problems for 99,9% of cases
**But you need to remove default value for allowed characters**
- existing sites that do not have latin-only limitation for username, will start having this limitation, even if you do not add to PHP code, adding default value to XML file will cause this as soon as user resave com_users configuration
- plus it will cause problems with entering emails as username which is common case, but there is a way out of adding this see last note

so remove default value from XML file and **inside your code check if parameter is empty** because now you do not check if it is empty, if it is empty do not try to enforce any character limitation

-

- [ ] Terminology,

 you have a parameter "charset" which has a name that is not appopriate: **charset** which usually implies the **name** of a character set aka values 'utf-8'  , 'utf-16' etc, so better call parameter:  "allowed_chars" and because it concerns the username, the parameter name should be:
**allowed_chars_username**

-

- [ ] Easy configuration of common cases for allowed characters without typing / avoiding problematic default values

since you can not have default value as explained above, but we don't want users to type all allowed characters and even make mistakes,
- the user should be to offered common cases via parameter, "allowed characters" parameter, for start only add these cases:
```
<field name="allowed_chars_username_preset" type="list"  defaut="0" ... >
  <option value="0">No limit</option>
  <option value="1">- Custom - </option>
  <option value="2">Latin Only</option>
  <option value="3">Email valid characters</option>
</field>
```

- for case 0 "- Custom -" you will use the characters from the textarea
- for latin only set allowed characters directly in your code
- to validate email case please use this code (**someone please correct me**)

```
$v = $value;

// This cleans some of the more dangerous characters but leaves special characters that are valid.
$noHtmlFilter = JFilterInput::getInstance();
$v = trim($noHtmlFilter->clean($v, 'HTML'));

// <>" are never valid in a email ?
$v = str_replace(array('<', '>', '"'), '', $v);

// Convert to Punycode string
$v = JStringPunycode::emailToPunycode( $v );

// Check for valid email (punycode is ASCII so this should work with UTF-8 too)
$email_regexp = "/^[a-zA-Z0-9.!#$%&'*+\/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*$/";

if (!preg_match($email_regexp, $v))
{
   // SET ERROR message and return false
}
```

Bakual commented:
Changed the title of the PR to reflect that it's no longer a plugin.

smz commented:
@ggppdk 
> to validate email case please use this code (someone please correct me)

We already have `JMailHelper::isEmailAddress()` (_amongst others..._) to validate email addresses. Please do not reinvent the wheel! :smile:

framontb committed:
Terminology change: charset for allowed_chars_username

framontb committed:
Added new tab for Password options in com_user.

framontb committed:
Preset charsets: configuration of common cases

framontb committed:
Enforcing coding standards by phpcs.

framontb committed:
No more Line exceeds 150 characters;

framontb commented:
Why the status is **failure** here while the status is "**All checks have passed**" in [github](https://github.com/joomla/joomla-cms/pull/9943 "github")?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9943">issues.joomla.org/joomla-cms/9943</a>.</sub>

mannybiker commented:
Very useful feature framontb! I was looking for this feature and found your PR right on time :)
I think it is correct to insert this as a core functionality. Username with spaces and multiple spaces are correctly considered different usernames, but this leads to problems most of the times in the real life and I think an admin should have the possibility to limit this behaviour as a core option, since usernames registration is a sensible and primordial part in a CMS.

I have tested your PR with Standard and Custom character set and it seems good to me.

framontb committed:
Added check for latin only character set.

framontb committed:
phpcs found a whitespace at end of line

framontb committed:
Added check for CUSTOM FORBIDDEN character set.

framontb committed:
Added check for CUSTOM REGEXP character set.

framontb committed:
phpcs: Cast statements must be followed by a single space

framontb committed:
Added check for M3AAWG East Asian languages

framontb commented:
**Insight request**:
Added to "username check" two new features:
First, give chance to administrator for supplying his own regexp, for restricting username field. So the admin could do whatever he wants with username field, in an absolutely custom way. Too far?
Second, implement "M3AAWG highly restrictive definitions for East Asian languages" from docs supplied by @chrisdavenport. I write somo regexps, but I think it will be worth if someone check the job. The M3AAWG doc is intended to support this combinations (if I understood well):
● Latin + Han + Hiragana +Katakana;
● Latin + Han + Bopomofo; 
● Latin + Han + Hangul;
Please if someone could have a look...

smz commented:
Should this be tested starting from the "staging" or the "3.6.x" branch? I see things which I don't understand, like an "_Integration_" tab **with untranslated strings** and this seems having to do with @Hackwar new router...

smz commented:
@framontb 

If I understand it correctly the M3AAWG document you cited in the code comment (https://www.m3aawg.org/sites/default/files/m3aawg-unicode-tutorial-2016-02.pdf ) says that for each identifier (_the "user name" in our case_) all characters should come from **a single script** (_e.g. Latin **or** Greek **or** Cyrillic_) **or a selected combination of scripts** (_e.g.: the three ones you cited_).

I haven't tested yet, but TBH and at first sight it doesn't seems to me that your `case 6:` is doing that, but I may be wrong.

framontb commented:
@smz 
The PR I sent is for 3.6.x.
Do you know a way to detect if all characters come from a single script? A simple way seems to check against each one of them... but how many of them are there? You quoted three (e.g. Latin or Greek or Cyrillic), are these enough for the purposes? Sorry, about so many questions.

smz commented:
@framontb 
> Do you know a way to detect if all characters come from a single script?

No, I don't. It is quite complicated stuff if you think that Unicode is formed by 17 "_planes_" each of 2^16 code points, of those 17 planes 3 are allocated to "Basic", "Supplemental" and "Ideographic" multilingual characters and just the "Basic Multilingual Plane" is divided in 160 blocks (_of different sizes, IIRC_), each roughly corresponding to a "_script_". And I don't know about the other 2 multilingual planes...

framontb commented:
Difficult issue for me, posted a question in stackoverflow :
[PHP regex for checking if a string comes from single script whatever it was](http://stackoverflow.com/questions/36984891/php-regex-for-checking-if-a-string-comes-from-single-script-whatever-it-was)
The trivial thing is check against each one. We will see if someone gives better idea.

framontb committed:
UNICODE Security MECHANISMS IMPLEMENTATION of Highly Restrictive level

framontb committed:
Disallow UAX31-D1. Default Identifier Syntax: <identifier> := <ID_Start> <ID_Continue>*

framontb commented:
After some research, I get an implementation of the **Unicode security mechanism: Highly Restrictive level**, first introduced in the M3AAWG docs above.
I think it deserve some explanation here because of the regex. 
This implementation, follows guidelines in [UNICODE SECURITY MECHANISMS](http://www.unicode.org/reports/tr39/). The restrictions applied for username field are:

1. Whenever **scripts** are tested for in the following definitions, characters with Script_Extension=Common and Script_Extension=Inherited are ignored.

2. Only some scripts are recommended for Highly Restrictive level. These recommended scripts are specified [here.](http://www.unicode.org/reports/tr31/#Table_Recommended_Scripts). I left the 'aspirational scripts' out. If someone think, it should be in, say.

3. All characters in each identifier must be from a single script, or from the [combinations](http://www.unicode.org/reports/tr39/#Restriction_Level_Detection).

4. No characters in the identifier (username) can be outside of the Identifier Profile. Because of this restriction, there is a big regex in the script ($regexAllowedChars). Although long, the structure of this regex is very simple, a list of characters (ranges) allowed. I find no other way to access this Identifier Profile that a container regex.

5. Finally there is a commented implementation of the [Default Identifier Syntax.](http://www.unicode.org/reports/tr31/#Default_Identifier_Syntax). I commented this because it seems to me very restrictive for the field username, disallowing all the unicode characters that no fit `/[^\p{Lu}\p{Ll}\p{Lt}\p{Lm}\p{Lo}\p{Nl}\p{Mn}\p{Mc}\p{Nd}\p{Pc}]/u`

Comments are welcome.

smz commented:
@framontb 
As you know I'm essentially against this PR, but I find this last work you did for implementing the Unicode security mechanism really commendable and possibly useful in many other circumstances. :+1: 

So much that I propose to move that code in an ad-hoc method for general usage. I don't know which library would be the best for hosting it, but I really hope the PLT could find a place for it.

Bravo! :clap:

smz commented:
... actually this, after implementing the necessary options for selecting the various possible profiles, could even be proposed as a PHP function

framontb commented:
It would be fine such a command (in PHP or Joomla!), because this check would be arranged with only one line of code, like in the email case.
I found this interesting tool in inet: [Homoglyph Attack Generator](http://www.irongeek.com/homoglyph-attack-generator.php). Simply, you put your username and choose the most similar chars.
Could be Joomla! the first CMS that implements this security check?

zero-24 commented:
Can we get a `New Feature` label here too? @brianteeman @wilsonge

brianteeman commented:
Please also take a look at a previous issue #7158 that the rules for usernames in the installation of joomla (Super Users) is not the same as that for users created in Joomla. Would be great of we could gt both fixed at the same time<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9943">issues.joomla.org/joomla-cms/9943</a>.</sub>

framontb committed:
Inadequate filtering of user-name during installation process. Issue #7158 , Issue #10391

framontb commented:
Fixed Issue #7158, # 10391 in previous commit.

manojLondhe commented on administrator/language/en-GB/en-GB.com_users.ini:
```diff
@@ -65,8 +65,26 @@ COM_USERS_CONFIG_FIELD_USERACTIVATION_DESC="If set to None the user will be regi
 COM_USERS_CONFIG_FIELD_USERACTIVATION_OPTION_SELFACTIVATION="Self"
+COM_USERS_CONFIG_FIELD_USERNAME_CHARSPRESET_LABEL="Characters preset"
+COM_USERS_CONFIG_FIELD_USERNAME_CHARSPRESET_DESC="Specify from the select list if you want disable this check or allow/forbid CUSTOM characters filling the CUSTOM text box. You can also select some of the specified character presettings: alphanumeric, latin, email,..."
+COM_USERS_CONFIG_FIELD_USERNAME_MINNUMCHARS_LABEL="MINimum number of characters"
```
@framontb 
Hi, can these  `MINimum`, `MAXimum`, `CUSTOM`be changed to `Minimum`, `Maximum`, `Custom`
Thanks for PR :)

Devportobello commented:
Dunno if this is an issue but "username rule" depend of the "com_users language ini", so if we use this rule outside of the com_users component, there is not translation... noticed the same for "password rule"

conconnl commented:
First I need to say this is a great improvement idea.
Maybe it's good (if possible) to add a email address blacklist feature for users registering, so you could easily block some domain names with wildcard options.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9943">issues.joomla.org/joomla-cms/9943</a>.</sub>

framontb commented:
@manojLondhe, @Devportobello, @conconnl
Thanks for the comments ! ￼However it is unclear whether this PR is appropriate for Joomla. It seems that is under discussion (don't know where or who). It makes no sense to continue working on it until it becomes clear.

roland-d commented:
@framontb Thank you for working on this. I am in favor of including this into the core. Especially if the wildcard option is included, this way we can have a list of forbidden usernames as well.

Pinging @chrisdavenport to see if he is interested in including this for 3.7. What do you say Chris?

chrisdavenport commented:
@roland-d @framontb I think this will be a valuable addition to Joomla and I'd be keen to include it in 3.7.  From a quick scan of the code I think it still needs a bit of work, but I like where it's going.

brianteeman commented:
As its mentioned above we now have anRTC pr to move passwords to their own tab #10941

framontb commented:
With new RTC #10941 , seems this PR is over. If no one is against, I close it.

conconnl commented:
The other PR is only visual changes, your is about new functionalities.
I would hope you go further with it and extend it a little more so it can ultimately go in 3.7

roland-d commented:
@framontb Issue #10941 is only a cosmetic change moving the password options to it's own tab. So it is unrelated to this PR.

framontb commented:
Better strategy is to chop this PR in pieces to test it properly like Brian did because the code still needs a bit of work, perhaps from people with more experience.
